src/core/symbology-ng/qgsarrowsymbollayer.h

```diff
@@ -0,0 +1,159 @@
+    void setArrowStartWidth( double w ) { mArrowStartWidth = w; }
+    /** Get the unit for the arrow start width */
+    QgsSymbolV2::OutputUnit arrowStartWidthUnit() const { return mArrowStartWidthUnit; }
+    /** Set the unit for the arrow start width */
+    void setArrowStartWidthUnit( QgsSymbolV2::OutputUnit u ) { mArrowStartWidthUnit = u; }
+    /** Get the scale for the arrow start width */
+    QgsMapUnitScale arrowStartWidthUnitScale() const { return mArrowStartWidthUnitScale; }
+    /** Set the scale for the arrow start width */
+    void setArrowStartWidthUnitScale( const QgsMapUnitScale& s ) { mArrowStartWidthUnitScale = s; }
+
+    /** Get the current arrow head size */
+    double headSize() const { return mHeadSize; }
+    /** Set the arrow head size */
+    void setHeadSize( double s ) { mHeadSize = s; }
+    /** Get the unit for the head size */
+    QgsSymbolV2::OutputUnit headSizeUnit() const { return mHeadSizeUnit; }
+    /** Set the unit for the head size */
+    void setHeadSizeUnit( QgsSymbolV2::OutputUnit u ) { mHeadSizeUnit = u; }
+    /** Get the scale for the head size */
+    QgsMapUnitScale headSizeUnitScale() const { return mHeadSizeUnitScale; }
+    /** Set the scale for the head size */
+    void setHeadSizeUnitScale( const QgsMapUnitScale& s ) { mHeadSizeUnitScale = s; }
+
+    /** Return whether it is a curved arrow or a straight one */
+    bool isCurved() const { return mIsCurved; }
+    /** Set whether it is a curved arrow or a straight one */
+    void setIsCurved( bool isCurved ) { mIsCurved = isCurved; }
+
+    /** Possible head types */
+    enum HeadType
+    {
+      HeadSingle,   //< One single head at the end
+      HeadReversed, //< One single head at the beginning
+      HeadDouble    //< Two heads
+    };
+
+    /** Get the current head type */
+    HeadType headType() const { return mHeadType; }
+    /** Set the head type */
+    void setHeadType( HeadType t ) { mHeadType = t; }
+
+    /**
+     * Should be reimplemented by subclasses to return a string map that
+     * contains the configuration information for the symbol layer. This
+     * is used to serialize a symbol layer perstistently.
+     */
+    QgsStringMap properties() const override;
+
+    /**
+     * Returns a string that represents this layer type. Used for serialization.
+     * Should match with the string used to register this symbol layer in the registry.
+     */
+    QString layerType() const override;
+
+    /** Prepare the rendering */
+    void startRender( QgsSymbolV2RenderContext& context ) override;
+
+    /** End of the rendering */
+    void stopRender( QgsSymbolV2RenderContext& context ) override;
+
+    /** Main drawing method */
+    void renderPolyline( const QPolygonF& points, QgsSymbolV2RenderContext& context ) override;
+
+  private:
+    /** Filling sub symbol. This is a shared pointer only because we need a unique pointer that is copyable (movable)*/
+    QSharedPointer<QgsFillSymbolV2> mSymbol;
```

Comment:
I'm a little confused - why does this need to be a shared pointer? I think it would be more expected behaviour if ever setSymbol creates a clone of the symbol or transfers ownership of it. Otherwise there could potentially be issues if setSymbol( some_symbol ) is called and the some_symbol is later modified.